lib/puppet/util/posix.rb

```diff
@@ -1,5 +1,48 @@
 # Utility methods for interacting with POSIX objects; mostly user and group
 module Puppet::Util::POSIX
+  require 'ffi'
+  module GrpLibC
+    extend FFI::Library
+    ffi_lib FFI::Library::LIBC
+    begin
+      attach_function :getgrouplist, [:string, :int, :pointer, :pointer], :int
+    rescue FFI::NotFoundError => e
+      puts e.message
+    end
+
+    def self.user_groups(user)
+      gid = Puppet::Etc.getpwnam(user).gid
+      ngroups_ptr = FFI::MemoryPointer.new(:int)
+      ngroups = 16
+      ngroups_ptr.write_int(ngroups)
+      groups_ptr = FFI::MemoryPointer.new(:uint, ngroups)
```

Comment:
Ruby will automatically free the memory when the pointer goes out of scope, but I like to make that explicit using the block form

```ruby 
FFI::MemoryPointer.new(:unit, ngroups) { |groups_ptr| ...  }
```